README.md
## How to test:

### 🔥 Pytest:

- https://docs.pytest.org/en/stable/usage.html
- https://docs.pytest.org/en/stable/example/index.html
- pytest-datafiles: https://github.com/omarkohl/pytest-datafiles/blob/master/README.rst

> For running all tests write down:

```bash
$ pytest [-q | -v]
```

> For running some specific tests (Ex: sorting):

```bash
$ python3 -m pytest ads/tests/sortings.py
$ pytest -k sorting.py
$ pytest ads/test/sortings.py
$ pytest -k -s test_avl_tree   # Use -s flag to allow to execute print statement in your test functions.
```

## How to Format:

### 🔥 Black:

- https://black.readthedocs.io/en/stable/compatible_configs.html

For formatting the whole package

```bash
$ python -m black {source_file_or_directory}
```

For formatting a module from it's diractory

```bash
$ black {source_file_or_directory}
```

# Implementaed Data Structures:

- [AVLTree](src/avl_tree.py)


how_to_markdown/changelog.md
# CHANGELOG


![📦 NEW:](https://img.shields.io/badge/-NEW-gray.svg?colorB=3778FF)
![📦 OLD:](https://img.shields.io/badge/Hellooo-MyMessage-orange)

All notable changes in this project's [released versions](https://github.com/ahmadawais/shades-of-purple-vscode/blob/master/../../releases) are documented in this file.

> [**Support my work**][sponsor]: If you or your company use any of my projects or like what I’m doing then consider [backing me][sponsor]. I'm in this for the long run. An open-source developer advocate.

[![Sponsor Awais](https://img.shields.io/badge/-Sponsor%20Awais%20%E2%86%92-gray.svg?colorA=6A788D&colorB=6A788D&style=flat)](https://github.com/AhmadAwais/sponsor/?utm_source=FOSS) [![Node.js CLI Automation](https://img.shields.io/badge/-NodeCLI.com%20%E2%86%92-gray.svg?colorA=6A788D&colorB=6A788D&style=flat)](https://NodeCLI.com/?utm_source=FOSS)
[![VSCode](https://img.shields.io/badge/-VSCode.pro%20%E2%86%92-gray.svg?colorA=6A788D&colorB=6A788D&style=flat)](https://VSCode.pro/?utm_source=GitHubFOSS)
[![Follow @MrAhmadAwais on Twitter](https://img.shields.io/twitter/follow/mrahmadawais.svg?style=social&label=Follow%20@MrAhmadAwais)](https://twitter.com/mrahmadawais/)

<br>

I have released a video course to help you build <a href="https://NodeCLI.com/?utm_source=FOSS" target="_blank">Node.js CLIs for Automation</a> →</p>

<a href="https://NodeCLI.com/?utm_source=FOSS" target="_blank"><img src="https://raw.githubusercontent.com/ahmadawais/stuff/master/nodecli/featured.jpg" /><br>NodeCLI.com</a> →

<br>

[![hr](https://raw.githubusercontent.com/ahmadawais/stuff/master/images/git/hr.png)](https://github.com/ahmadawais/shades-of-purple-vscode/blob/master/)

**CHANGELOG KEY**: `📦 NEW`, `👌 IMPROVE`, `🐛 FIX`, `📖 DOC`, `🚀 RELEASE`, `🤖 TEST`, and `‼️ BREAKING`.

<small>

> _I use [Emoji-log](https://github.com/ahmadawais/Emoji-Log), you should try it and simplify your git commits._

</small>

<br>

[![hr](https://raw.githubusercontent.com/ahmadawais/stuff/master/images/git/hr.png)](https://github.com/ahmadawais/shades-of-purple-vscode/blob/master/)

<br>

### RELEASE: [v6.12.0](https://github.com/ahmadawais/shades-of-purple-vscode/compare/6.11.0...v6.12.0)

![📦 NEW:](https://img.shields.io/badge/-NEW-gray.svg?colorB=3778FF)
![📦 OLD:](https://img.shields.io/badge/Hellooo-MyMessage-orange)

> 📦 tab state colors [`c1d0457`](https://github.com/ahmadawais/shades-of-purple-vscode/commit/c1d04571c941c980295af5c7727140c4676e8dc1) <br>

![📖 DOC:](https://img.shields.io/badge/-DOCS-gray.svg?colorB=978CD4)

> 📖 Alacritty [`8d4b591`](https://github.com/ahmadawais/shades-of-purple-vscode/commit/8d4b591149811b2bee166512032e896ca59849a3) <br>

<br>

[![hr](https://raw.githubusercontent.com/ahmadawais/stuff/master/images/git/hr.png)](https://github.com/ahmadawais/shades-of-purple-vscode/blob/master/)

<br>

### RELEASE: [6.11.0](https://github.com/ahmadawais/shades-of-purple-vscode/compare/6.10.0...6.11.0)

![🐛 FIX:](https://img.shields.io/badge/-FIX-gray.svg?colorB=ff6347)

> 🐛 TypeScript Variables [`ea87077`](https://github.com/ahmadawais/shades-of-purple-vscode/commit/ea87077b722cd2eff93b27b3b85e112835380aa0) <br>

![📖 DOC:](https://img.shields.io/badge/-DOCS-gray.svg?colorB=978CD4)

> 📖 Changelog [`c7720a3`](https://github.com/ahmadawais/shades-of-purple-vscode/commit/c7720a33b6af63bc36ddeb7a2021087e80043a16) <br>

<br>

[![hr](https://raw.githubusercontent.com/ahmadawais/stuff/master/images/git/hr.png)](https://github.com/ahmadawais/shades-of-purple-vscode/blob/master/)

<br>

### RELEASE: [6.10.0](https://github.com/ahmadawais/shades-of-purple-vscode/compare/6.9.0...6.10.0)

![👌 IMPROVE:](https://img.shields.io/badge/-IMPROVEMENT-gray.svg?colorB=39AA54)

> 👌 Typescript Syntax [`651b8ad`](https://github.com/ahmadawais/shades-of-purple-vscode/commit/651b8ad823c9a0e6b8aba2e81b9aa996aa754e3b) <br>

![📖 DOC:](https://img.shields.io/badge/-DOCS-gray.svg?colorB=978CD4)

> 📖 Changelog [`5d8f64d`](https://github.com/ahmadawais/shades-of-purple-vscode/commit/5d8f64df02ca18107d6f1a9cbc035966f0925ba7) <br>

<br>

[![hr](https://raw.githubusercontent.com/ahmadawais/stuff/master/images/git/hr.png)](https://github.com/ahmadawais/shades-of-purple-vscode/blob/master/)

<br>

### RELEASE: [6.9.0](https://github.com/ahmadawais/shades-of-purple-vscode/compare/6.8.0...6.9.0)

![📦 NEW:](https://img.shields.io/badge/-NEW-gray.svg?colorB=3778FF)

> 📦 Badge [`e084795`](https://github.com/ahmadawais/shades-of-purple-vscode/commit/e08479588ec09a170e54553e4e2e09a04018ac11) <br>

![📖 DOC:](https://img.shields.io/badge/-DOCS-gray.svg?colorB=978CD4)

> 📖 Changelog [`df6f582`](https://github.com/ahmadawais/shades-of-purple-vscode/commit/df6f582590b900caa17aa508a2844502d0aa55e9) <br>

<br>

[![hr](https://raw.githubusercontent.com/ahmadawais/stuff/master/images/git/hr.png)](https://github.com/ahmadawais/shades-of-purple-vscode/blob/master/)

<br>

### RELEASE: [6.8.0](https://github.com/ahmadawais/shades-of-purple-vscode/compare/6.7.0...6.8.0)

![👌 IMPROVE:](https://img.shields.io/badge/-IMPROVEMENT-gray.svg?colorB=39AA54)

> 👌 Remove Extraneous Dev [`e82d708`](https://github.com/ahmadawais/shades-of-purple-vscode/commit/e82d7085b4088f6400838683fe3817811f5bf98b) <br>

![📖 DOC:](https://img.shields.io/badge/-DOCS-gray.svg?colorB=978CD4)

> 📖 Changelog [`b4537ad`](https://github.com/ahmadawais/shades-of-purple-vscode/commit/b4537ad3e053a04bfb5aa266557ff9313d323b19) <br>

<br>

[![hr](https://raw.githubusercontent.com/ahmadawais/stuff/master/images/git/hr.png)](https://github.com/ahmadawais/shades-of-purple-vscode/blob/master/)

<br>

### RELEASE: [6.7.0](https://github.com/ahmadawais/shades-of-purple-vscode/compare/6.6.0...6.7.0)

![👌 IMPROVE:](https://img.shields.io/badge/-IMPROVEMENT-gray.svg?colorB=39AA54)

> 👌 Numbers [`342af22`](https://github.com/ahmadawais/shades-of-purple-vscode/commit/342af22c6da7d7696852219d85f5cbfe81241348) <br>

![📖 DOC:](https://img.shields.io/badge/-DOCS-gray.svg?colorB=978CD4)


test/test_avl_tree.py
import os
import pytest
import unittest

from utils import *
from avl_tree import *




class TestAVLTreeST(unittest.TestCase):
    @classmethod
    def setUpClass(self):
        data = "Algorithms_and_Data_Structures/ads/data/shellsST.txt"
        self.DATA_FILE_PATH = os.path.abspath(os.path.join(os.pardir, data))
        print("\nTesting TestAVLTreeST ......")

    def setUp(self):
        self.kvp = NUMBERS

        self.T1 = AVLTreeST()
        self.T2 = AVLTreeST()

        load_data_from_file(self.DATA_FILE_PATH, self.T1)
        load_data_from_collection(self.kvp, self.T2)

    def tearDown(self):
        for key in self.T1.keys():
            self.T1.delete(key)
        assert self.T1.is_empty()
        self.T1 = None

    def test_size(self):
        self.assertEqual(self.T1.size(), 7)
        self.assertEqual(self.T2.size(), len(self.kvp))

    def test_is_empty(self):
        self.assertFalse(self.T1.is_empty())
        self.assertFalse(self.T2.is_empty())

    def test_height(self):
        self.assertEqual(self.T2.height(), 2)

        self.T2.delete("ZERO")
        self.T2.delete("TEN")
        self.T2.delete("TWENTY")
        self.assertEqual(self.T2.height(), 1)

        self.T2.delete("THIRTY")
        self.T2.delete("FORTY")
        self.assertEqual(self.T2.height(), 0)

        self.T2.delete("FIFTY")
        self.assertEqual(self.T2.height(), -1)



    def test_contains(self):
        self.assertIn("by", self.T1)
        self.assertIn("she", self.T1)
        self.assertNotIn("bye", self.T1)  # ≠ self.assertFalse('bye', self.T1)

        self.assertNotIn("APPLE", self.T2)  # ≠ self.assertFalse(1003, self.T2)

        self.assertIn("ZERO", self.T2)
        self.assertIn("FIFTY", self.T2)

        with self.assertRaises(ValueError):
            self.T2.contains(None)
            self.T1.contains(None)

    def test_get(self):
        self.assertEqual(self.T1.get("by"), 4)
        self.assertEqual(self.T1.get("she"), 0)

        self.assertIsNone(self.T2.get("APPLE"))

        self.assertEqual(self.T2.get("ZERO"), 0)
        self.assertEqual(self.T2.get("THIRTY"), 30)
        self.assertEqual(self.T2.get("TWENTY"), 20)
        self.assertEqual(self.T2.get("FIFTY"), 50)

        with self.assertRaises(ValueError):
            self.T1.get(None)
            self.T2.get(None)

    def test_put(self):
        self.T2.put("EIGHTY", 80)
        self.assertIn("EIGHTY", self.T2)
        self.assertEqual(self.T2.get("EIGHTY"), 80)
        self.assertEqual(self.T2.size(), len(self.kvp) + 1)

        with self.assertRaises(ValueError):
            self.T1.put(None, randrange(0, 11))
            self.T2.put(None, randrange(0, 11))

        self.T2.delete("EIGHTY")

    def test_delete(self):
        self.T2.delete("ZERO")
        self.assertNotIn("ZERO", self.T2)
        self.assertEqual(self.T2.size(), len(self.kvp) - 1)

        self.T2.delete("THIRTY")
        self.assertNotIn("THIRTY", self.T2)
        self.assertEqual(self.T2.size(), len(self.kvp) - 2)

        with self.assertRaises(ValueError):
            self.T1.delete(None)
            self.T2.delete(None)

        self.T2.put("ZERO", 0)
        self.T2.put("THIRTY", 30)

    def test_select(self):
        self.assertEqual(self.T2.select(0), "FIFTY")
        self.assertEqual(self.T2.select(3), "THIRTY")

        T3 = AVLTreeST()
        with self.assertRaises(ValueError):
            self.T2.select(-1)
            self.T2.select(8)
            T3.select(-1)

    def test_rank(self):
        self.assertEqual(self.T2.rank("APPLE"), 0)  # key not in tree
        self.assertEqual(self.T2.rank("FIFTY"), 0)
        self.assertEqual(self.T2.rank("FORTY"), 1)
        self.assertEqual(self.T2.rank("ZERO"), len(self.kvp) - 1)
        self.assertEqual(self.T2.rank("ZEROO"), len(self.kvp))  # key not in tree
        self.assertEqual(self.T2.rank("ZZERO"), len(self.kvp))  # key not in tree

        T3 = AVLTreeST()
        with self.assertRaises(ValueError):
            self.T2.rank(None)
            T3.rank(None)

    def test_keys_inrange(self):
        # key not in tree
        # self.assertEqual(self.T2.keys_inrange(10001, 10029), 7)
        self.assertEqual(self.T2.keys_inrange("APPLE", "ZERO"), [])
        self.assertEqual(self.T2.keys_inrange("FIFTY", "FIFTY"), ["FIFTY"])
        self.assertEqual(self.T2.keys_inrange("FIFTY", "FORTY"), ["FIFTY"])
        self.assertEqual(self.T2.keys_inrange("FORTY", "FIFTY"), [])

        T3 = AVLTreeST()
        with self.assertRaises(ValueError):
            self.T2.keys_inrange(None, None)
            T3.keys_inrange(10001, None)
            T3.keys_inrange(None, 10003)

    def test_size_inrange(self):
        self.assertEqual(self.T2.size_inrange("FIFTY", "ZERO"), len(self.kvp))
        self.assertEqual(self.T2.size_inrange("APPLE", "FIFTY"), 1)
        self.assertEqual(self.T2.size_inrange("FIFTY", "APPLE"), 0)
        self.assertEqual(self.T2.size_inrange("ZERO", "ZERO"), 1)
        self.assertEqual(self.T2.size_inrange("ZERO", "ZZERO"), 1)
        self.assertEqual(self.T2.size_inrange("TWENTY", "ZERO"), 2)

        T3 = AVLTreeST()
        with self.assertRaises(ValueError):
            self.T2.size_inrange(None, None)
            T3.size_inrange(10001, None)

    def test_keys(self):
        keys_inorder = [key for key in self.kvp.keys()]
        keys_inorder.sort()
        self.assertEqual(self.T2.keys(), keys_inorder)

    def test_keys_level_order(self):
        keys = ["TEN", "FORTY", "TWENTY", "FIFTY", "THIRTY", "ZERO"]
        self.assertListEqual(self.T2.keys_level_order(), keys)

    def test_checked(self):
        self.assertTrue(self.T1.checked())
        self.assertTrue(self.T2.checked())

        T3 = AVLTreeST()
        self.assertTrue(T3.checked())

    def test_is_AVL(self):
        self.assertTrue(self.T1.is_AVL())
        self.assertTrue(self.T2.is_AVL())

        T3 = AVLTreeST()
        self.assertTrue(T3.is_AVL())

    def test_is_BST(self):
        self.assertTrue(self.T1.is_BST())
        self.assertTrue(self.T2.is_BST())

        T3 = AVLTreeST()
        self.assertTrue(T3.is_BST())

    def test_is_size_consistent(self):
        self.assertTrue(self.T1.is_size_consistent())
        self.assertTrue(self.T2.is_size_consistent())

        T3 = AVLTreeST()
        self.assertTrue(T3.is_size_consistent())

    def test_is_rank_consistent(self):
        self.assertTrue(self.T1.is_rank_consistent())
        self.assertTrue(self.T2.is_rank_consistent())

        T3 = AVLTreeST()
        self.assertTrue(T3.is_rank_consistent())


def simple_test_AVLTreeST():
    ST = AVLTreeST()
    load_data_from_collection(NUMBERS, ST)

    print('*'*25, 'Before Mutation: ', '*'*25)
    display_st(ST)

    del ST["ZERO"]
    del ST["FIFTY"]
    ST.put("NINTY", 90)
    ST['SIXTY'] = 60
    del ST['TWENTY']

    print('*'*25, 'After Mutation: ', '*'*25)
    display_st(ST)


test/test_sortings.py
import unittest
from math import floor, sqrt
from random import randint, randrange, choice

from .utils import *
from .sortings import *
import pytest



class TestSorting(unittest.TestCase):

    def setUp(self):
        _NUM = 1000
        _sorted = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
        _repeating_sorted = [-10, -10, -10, 0, 0, 0, 0, 1, 8, 8, 9, 9, 9]
        _reverse = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
        _not_sorted = [15, 8, 5, 12, 17, 18, 4, 13, 14, 19]
        _equal = [1, 1, 1, 1, 1, 1, 1, 1]
        _negative = [-1, 0, 5, -10, 20, 13, -7, 3, 2, -3]
        _negative_sorted = [-10, -7, -3, -1, 0, 2, 3, 5, 13, 20]

        self.list_of_lists = [_sorted, _repeating_sorted, _reverse,
                              _not_sorted, _equal, _negative, _negative_sorted]
        
        self.list_of_lists_B = [
            [randrange(0, _NUM) for x in range(_NUM)] for i in range(floor(sqrt(_NUM)))
        ]

    def tearDown(self):
        pass

# #************************* Insertion Sort Testings: **************************#
#     def test_insertion_sort(self):
#         for array in self.list_of_lists:
#             insertion_sort(array)
#             assert is_sorted(array)

# #************************* Selection Sort Testings: **************************#
#     def test_selection_sort(self):
#         for array in self.list_of_lists:
#             selection_sort(array)
#             assert is_sorted(array)

# #************************* Bubble Sort Testings: **************************#
    # def test_bubble_sort():
    #     my_list = [12, 5, 13, 8, 9, 65]
    #     bubble_sort(my_list)
    #     print(my_list)


#*************************** Mergesort Testings: *****************************#
    def test_merge_sort(self):

        for array in self.list_of_lists:
            L = merge_sort(array)
            assert is_sorted(L)

    def test_merge_sort_two(self):

        for array in self.list_of_lists:
            merge_sort_two(array)
            assert is_sorted(array)

    def test_merge_sort_three(self):

        for array in self.list_of_lists:
            merge_sort_three(array)
            assert is_sorted(array)

#*************************** QuickSort Testings: *****************************#

    def test_quick_sort(self):
        for array in self.list_of_lists:
            quick_sort(array)
            assert is_sorted(array)

    def test_quick_sort_two(self):

        for array in self.list_of_lists:
            quick_sort_two(array)
            assert is_sorted(array)

        for array in self.list_of_lists_B:
            quick_sort_two(array)
            assert is_sorted(array)

_NUM = 1000
_sorted = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
_repeating_sorted = [-10, -10, -10, 0, 0, 0, 0, 1, 8, 8, 9, 9, 9]
_reverse = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
_not_sorted = [15, 8, 5, 12, 17, 18, 4, 13, 14, 19]
_equal = [1, 1, 1, 1, 1, 1, 1, 1]
_negative = [-1, 0, 5, -10, 20, 13, -7, 3, 2, -3]
_negative_sorted = [-10, -7, -3, -1, 0, 2, 3, 5, 13, 20]

list_of_lists = [_sorted, _repeating_sorted, _reverse, _not_sorted, _equal, _negative, _negative_sorted]
list_of_lists_B = [[randrange(0, _NUM) for x in range(_NUM)] for i in range(floor(sqrt(_NUM)))]


pytest.mark.parametrize(list_of_lists)
def test_quick_sort_two():

    for array in list_of_lists:
        quick_sort_two(array)
        assert is_sorted(array)


test/test_searching.py
from searching import *
import unittest
import pytest
from random import *

data = [[[*range(5)], 2, 2], [[5, 4, 3, 2], 5, 0],
        [[0], 1, -0.0], [[1, 3, 6, 9], 9, 3]]


@pytest.mark.parametrize("L, k, expected", data)
def test_search_first_of_k(L, k, expected):
    assert search_first_of_k(L, k) == expected


class TestBinarySearch(unittest.TestCase):

    @classmethod
    def setUpClass(cls):

        cls.L0 = [[[*range(5)], 2, 2], [[5, 4, 3, 2], 5, 0],
                  [[0], 1, -0.0], [[1, 3, 6, 9], 9, 3]]

    def setUp(self):
        r = randrange(0, 1000)
        self.L1 = [[[*range(5)], 2, 2], [[5, 4, 3, 2], 5, 0],
                   [[0], 1, -0.0], [[1, 3, 6, 9], 9, 3]]
        self.k = choice([*range(-10, 10)])

    # @pytest.mark.parametrize("L, k, expected", self.L1)
    def test_search_first_of_k(self, L, k, expected):
        assert search_first_of_k(L, k) == expected

    # def test_idx_k_rec(self):
    #     self.assertEqual(self.k, self.L[idx_k_rec(
    #         self.L, self.k, 0, len(self.L)-1)])

    # def test_idx_k_nearnest(self):
    #     self.assertEqual(self.k, self.L[idx_k_nearest(self.L, self.k)])

    # def test_idx_k_nearest_rec(self):
    #     self.assertEqual(self.k, self.L[idx_k_nearest_rec(
    #         self.L, self.k, 0, len(self.L)-1)])

    def tearDown(self):
        pass


####################################################################
####################################################################


src/sortings.py
from random import randint, randrange
from typing import List
import sys
from utils import *



"""Implement Selection sort. """


"""Implementation of Insertion Sort."""


"""Implementation of Bubble Sort."""

# NOT TESTED !!
def bubble_sort(L):
    for j in range(len(L)): 
        for i in range(len(L) - 1):
            if L[i] > L[i+1]:
                temp = L[i]
                L[i] = L[i+1]
                L[i+1] = temp


#--------------------------------------------------
"""Implementation of merge sort algorithm.

This modules provides several methods, implemented in various way, for
sorting a list using mergesort algorithm.
"""


def merge_sort(L: List) -> List:
    """Sorts and returns the specified list using merge-sort algorithm.

    Args:
        L: the list to be sorted
    Returns:
        a sorted list containing all elements of the given list, L
    """

    N = len(L)
    if N <= 1:
        return L

    mid = N // 2
    left = merge_sort(L[0: mid])
    right = merge_sort(L[mid:])
    return merge(left, right)


def merge(L1: List, L2: List) -> List:
    """It returns a sorted list after stably merging L1 with L2.

    Args:
        L1: sorted list to be merged with L2 into a new list
        L2: sorted list to be merged with L1 into a new list
    Returns:
        a sorted list containing all elements of both list - L1 and L2.
    """
    L = L1 + L2
    left_index, right_index = 0, 0

    for i in range(len(L)):
        if left_index >= len(L1):
            L[i] = L2[right_index]
            right_index += 1
        elif right_index >= len(L2):
            L[i] = L1[left_index]
            left_index += 1
        elif L2[right_index] < L1[left_index]:   # ensure stability
            L[i] = L2[right_index]
            right_index += 1
        else:
            L[i] = L1[left_index]
            left_index += 1
    return L


#************************ Alternative implementations ************************#
def merge_sort_two(a: List) -> None:
    """Sorts the  specified list using merge-sort algorithm.

    Args:
        a: the list to be sorted
    """
    _merge_sort_two(a, 0, len(a)-1)


def _merge_sort_two(a: List, low: int, high: int) -> None:
    """It sorts a[low ... high] by first deviding it into a[low ... mid] and
    a[mid+1 ... high] and then merging them

    Args:
        a   : the array to be sorted
        low : first index of first half of the array, a[low ... mid]
        high: last index of second half of the array, a[mid+1 ... high]
    """
    if low >= high:
        return

    mid = (low + high) // 2

    # sorts first half of the array, a[low ... mid]
    _merge_sort_two(a, low, mid)

    # sorts first half of the array, a[mid+1 ... high]
    _merge_sort_two(a, mid + 1, high)

    # merges a[low ... mid] and a[mid+1 ... high], two halves of the array
    merge_two(a, low, mid, high)


def merge_two(a: List, low: int, mid: int, high: int):
    """It stably merges a[low ... mid] with a[mid+1 ... high] using an
    auxilary array, aux[low ... high]

    Args:
        a   : the array is being sorted
        low : initial index of first half of the array, a[low ... mid]
        mid : last index of first half of the array, a[low ... mid]
        high: last index of second half of the array, a[mid+1 ... high]
    """

    aux = [item for item in a]
    left, right, i = low, mid+1, low

    while left <= mid and right <= high:
        if aux[left] <= aux[right]:       # ensures stability
            a[i] = aux[left]
            left += 1
            i += 1
        else:
            a[i] = aux[right]
            right += 1
            i += 1

    while left <= mid:
        a[i] = aux[left]
        i += 1
        left += 1


#************************ Alternative implementations ************************#

def merge_sort_three(a: List):
    """Sorts the  specified list using merge-sort algorithm.

    Args:
        a: the list to be sorted
    """
    aux = [None]*len(a)
    _merge_sort_three(a, 0, len(a)-1, aux)


def _merge_sort_three(a: List, low: int, high: int, aux: List) -> None:
    """It sorts a[low ... high] by first deviding it into a[low ... mid] and
    a[mid+1 ... high] and then merging them

    Args:
        a   : the array to be sorted
        low : first index of first half of the array, a[low ... mid]
        high: last index of second half of the array, a[mid+1 ... high]
        aux : the auxilary array for recycling
    """
    if low >= high:
        return
    mid = (low + high) // 2

    # sorts first half of the array, a[low ... mid]
    _merge_sort_three(a, low, mid, aux)

    # sorts first half of the array, a[mid+1 ... high]
    _merge_sort_three(a, mid + 1, high, aux)

    # merges a[low ... mid] and a[mid+1 ... high], two halves of a[low...high]
    merge_three(a, low, mid, high, aux)


def merge_three(a: List, low: int, mid: int, high: int, aux: List) -> None:
    """It stably merges a[low ... mid] with a[mid+1 ... high] using an
    auxilary array, aux[low ... high]

    Args:
        a   : the array is being sorted
        low : first index of first half of the array, a[low ... mid]
        mid : last index of first half of the array, a[low ... mid]
        high: last index of second half of the array, a[mid+1 ... high]
        aux : the auxilary array for recycling
    """
    for i in range(low, high+1):
        aux[i] = a[i]

    left_index, right_index = low, mid+1

    for i in range(low, high+1):
        if left_index > mid:
            a[i] = aux[right_index]
            left_index += 1
        elif right_index > high:
            a[i] = aux[left_index]
            right_index += 1
        elif aux[right_index] < aux[left_index]:   # ensures stability
            a[i] = aux[right_index]
            right_index += 1
        else:
            a[i] = aux[left_index]
            left_index += 1


""" Implementation of quick sort algorithm.

This modules provides two functions, implemented in different ways, for
sorting a list using quicksort algorithm.
"""


def quick_sort(a):

    def _quick_sort(a, low, high):
        if low >= high:
            return
        pi = partition(a, low, high)  # pi -> Pivot Index
        _quick_sort(a, low, pi-1)
        _quick_sort(a, pi+1, high)

    def partition(a, low, high):

        pivot, pi = a[high], low      # pi -> Pivot Index

        for i in range(low, high):
            if a[i] < pivot:
                swap(a, pi, i)
                pi = pi+1

        swap(a, pi, high)
        return pi

    # called the quicksort procedure
    _quick_sort(a, 0, len(a)-1)


def quick_sort_two(a):

    def _quick_sort_two(a, low, high):
        if low >= high:
            return
        pi = partition_two(a, low, high)  # pi -> Pivot Index
        _quick_sort_two(a, low, pi-1)
        _quick_sort_two(a, pi+1, high)

    def partition_two(a, low, high):
        """ partition the subarray a[low, ..., high].

        it partition the subarray a[low, ..., high] so that a[lo..j-1] <= a[j] <=
        a[j+1..hi] and return the index j.

        Args:
            a     -> the array to be partioned
            low   -> the lower index into the specified array
            hight -> the upper index into the specified array

        """
        pivot, i, j = a[low], low + 1, high
        while True:
            while a[i] < pivot:    # find item on low to swap
                if i == high:
                    break
                i += 1

            while a[j] >= pivot:    # find item on high to swap
                if j == low:
                    break  # redundant since a[low] acts as sentinel
                j -= 1

            if i >= j:
                break        # check if pointers cross

            swap(a, i, j)

        swap(a, low, j)             # put partitioning item pivot at a[j]

        # now, a[low ... j-1] <= a[j] <= a[j+1 ... high]
        return j

    # called the quicksort procedure
    _quick_sort_two(a, 0, len(a)-1)

"""Implementations of Least Significant Digit Radix Sort."""



"""Implementations of Most Significant Digit Radix Sort."""

# LSD -> Least Significent Digit
def LSD_radix_sort(a: list, word_len: int):
    """
    Args:
        a: list of stirng of SAME length.
        w: length of a item (string) in the list, a.
    """
    aux = [None]*len(a)
    RADIX = 256

    for w in reversed(range(word_len)):
        count = [0]*(RADIX+1)
        for i in range(len(a)):
            count[ord(a[i][w])+1] += 1
        for i in range(RADIX):
            count[i+1] += count[i]
        for i in range(len(a)):
            aux[count[ord(a[i][w])]] = a[i]
            count[ord(a[i][w])] += 1
        for i in range(len(a)):
            a[i] = aux[i]


src/utils.py
from typing import Sequence, List

NUMBERS = {"ZERO": 0, "TEN": 10, "TWENTY": 20, "THIRTY": 30, "FORTY": 40, "FIFTY": 50}

COLORS = sorted(["RED", "GREEN", "BLACK", "BLUE", "WHITE", "PINK", "YELLOW",
                 "CYAN", "MAGENTA", "GRAY", "BISQUE", "AZURE", "AQUA", "BEIGE"])


def swap(array: List[Sequence], i: int, j: int) -> None:
    tmp = array[i]
    array[i] = array[j]
    array[j] = tmp


def is_sorted(array):
    ''' Return True if the given list is sorted. '''
    for i in range(len(array)-1, 0, -1):
        if array[i] < array[i-1]:
            return False
    return True



def load_data_from_file(file_path, container):
    with open(file_path, encoding="utf-8") as f:         # Open the file

        value = 0
        while True:
            line = f.readline()
            if not line:
                break                      # line is empty so exit
            for word in line.split():
                container.put(word, value)
                value += 1


def load_data_from_collection(collection, container):

    if type(collection) is dict:
        for key in collection.keys():
            container.put(key, collection[key])

    elif type(collection) is list:
        for item in collection:
            container.append(collection[item])


def display_st(ST):
    print(f"""\t{"keys(): "} \t{ST.keys()}""")
    print(f"""\t{"root.key: "} \t{ST.root.key}""")
    print(f"""\t{"size(): "} \t{ST.size()}""")
    print(f"""\t{"height(): "} \t{ST.height()}""")
    print(f"""\t{"is_empty(): "} \t{ST.is_empty()}""")
    print(f"""\t{"contains('FORTY'): "} \t{ST.contains('FORTY')}""")
    print(f"""\t{"get('THIRTY'): "} \t{ST.get('THIRTY')}""")
    print(f"""\t{"select(3): "} \t{ST.select(3)}""")
    # print(f"""\t{"rank('THIRTY'): "} \t{ST.rank(ST.select(3))}""")
    print("\trank({}): \t{}".format(ST.select(3), ST.rank(ST.select(3))))
    print(f"""\t{"floor('TEEN'): "} \t{ST.floor('TEEN')}""")
    print(f"""\t{"ceiling('NINTYY'): "} \t{ST.ceiling('NINTYY')}""")

    print('\tkey-value pairs (IN-ORDER): ')
    for key in ST.keys():
        print(f"\t \t{key} \t{ST[key]}")


# FIXTURE_DIR = os.path.abspath(os.path.join(os.pardir, 'data'))


# @pytest.mark.datafiles(
#     os.path.join(FIXTURE_DIR, 'img1.jpg'),
#     os.path.join(FIXTURE_DIR, 'img2.jpg'),
#     os.path.join(FIXTURE_DIR, 'img3.jpg'),
# )
# def test_find_borders(datafiles):
#     for img in datafiles.listdir():
#         print(img)
        #assert process(img) == some_expected_value


# def simple_test_AVLTreeST(data):
#     ST = AVLTreeST()
#     load_data_from_collection(data, ST)

#     print('*'*25, 'Before Mutation: ', '*'*25)
#     display_st(ST)

#     del ST["ZERO"]
#     del ST["FIFTY"]
#     ST.put("NINTY", 90)
#     ST['SIXTY'] = 60
#     del ST['TWENTY']

#     print('*'*25, 'After Mutation: ', '*'*25)
#     display_st(ST)


# @pytest.mark.skip(reason='Not complete yet !')
# def test_main():

#     if len(sys.argv) > 1:
#         simple_test_AVLTreeST(sys.argv[1])
#     else:
#         simple_test_AVLTreeST(NUMBERS)
